cluster_utils/settings.py

```diff
@@ -314,74 +383,143 @@ def add_cmd_line_params(base_dict, extra_flags):
+        add_cmd_line_params(orig_dict, cmdline_settings)
+
+    return smart_settings.load(
+        os.fspath(settings_file),
+        make_immutable=make_immutable,
+        dynamic=dynamic,
+        post_unpack_hooks=([add_cmd_params, check_reserved_params] + post_unpack_hooks),
+        pre_unpack_hooks=pre_unpack_hooks,
+    )
+
+
+def read_params_from_cmdline(
+    cmd_line: Optional[list[str]] = None,
+    make_immutable: bool = True,
+    verbose: bool = True,
+    dynamic: bool = True,
+    save_params: bool = True,
+) -> smart_settings.AttributeDict:
+    """Read parameters based on command line input.
+
+    Args:
+        cmd_line:  Command line arguments (defaults to sys.argv).
+        make_immutable:  See ``smart_settings.loads()``
+        verbose:  If true, print the loaded parameters.
+        dynamic:  See ``smart_settings.loads()``
+        save_params:  If true, save the settings as JSON file in the working_dir.
+
+    Returns:
+        Parameters as loaded by smart_settings.
+    """
     if not cmd_line:
         cmd_line = sys.argv
 
-    try:
-        connection_details = ast.literal_eval(cmd_line[1])
-    except (SyntaxError, ValueError):
-        connection_details = {}
-        pass
-
-    if set(connection_details.keys()) == {constants.ID, "ip", "port"}:
-        submission_state.communication_server_ip = connection_details["ip"]
-        submission_state.communication_server_port = connection_details["port"]
-        submission_state.job_id = connection_details[constants.ID]
-        del cmd_line[1]
-        submission_state.connection_details_available = True
-        submission_state.connection_active = False
+    # expected keys of the server connection dictionary
+    server_connection_keys = {constants.ID, "ip", "port"}
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--server-connection-info",
+        type=ast.literal_eval,
+        help="""Information to communicate with the cluster_utils main process.
+           Dictionary with keys {}.
+       """.format(
+            server_connection_keys
+        ),
+    )
+    param_group = parser.add_mutually_exclusive_group(required=True)
+    param_group.add_argument(
+        "--parameter-dict",
+        type=ast.literal_eval,
+        help="Dictionary with the job parameters.",
+    )
+    param_group.add_argument(
+        "--parameter-file", type=pathlib.Path, help="File with the job parameters."
+    )
+    parser.add_argument(
+        "--parameters",
+        nargs="+",
+        type=str,
+        default=[],
+        metavar="KEY_VALUE",
+        help="""Additional parameters in the format '<key>=<value>'.  Values provided
+            here overwrite parameters provided via `--parameter-dict` or
+            `--parameter-file`.  Key has to match a configuration option, value has to
+            be valid Python.  Example: `--parameters 'results_dir="/tmp"'
+            'optimization_setting.run_local=True'`
+        """,
+    )
```

Comment:
I see your point.  I was struggling here, mostly because of the optional first positional argument (server info) in the original implementation, which is not really doable with argparse.
If it's okay to keep that one as named parameter (I assume server info is normally not needed when running manually?), changing the rest back to positionals should be doable.  I.e. the interface would then look like this
```
script.py [--server-connection-info ...] parameter_file_or_dict [key_value [key_value ...]]
```
which I think should be compatible with the old implementation as long as no server info is passed.

Reply:
Ah, one question here: Is it okay to make `parameter_file_or_dict` required?  With the old implementation, this was also optional, which makes it even more complicated.  I can't see a use-case for running without parameters but maybe I'm just missing something.

Reply:
I think I'd prefer the combined `parameter_file_or_dict`.  It's more argparse-friendly (there cannot be multiple optional positionals) and also avoids the potential case that someone passes both a file and a dict.

Reply:
I now came up with the following: `parameter_file_or_dict` is a combined argument as I suggested above but a separate flag `--dict` has to be passed when using a dictionary.  That is, there is no auto-detection of the type anymore, which makes it more explicit and helps to provide more useful error messages.